src/Databases/DatabaseAtomic.cpp

```diff
@@ -160,6 +160,46 @@ void DatabaseAtomic::dropTableImpl(ContextPtr local_context, const String & tabl
     DatabaseCatalog::instance().enqueueDroppedTableCleanup(table->getStorageID(), table, table_metadata_path_drop, sync);
 }
 
+void DatabaseAtomic::dropDetachedTable(ContextPtr local_context, const String & table_name, const bool sync)
+{
+    waitDatabaseStarted();
+
+    assert(fs::exists(getObjectMetadataPath(table_name)));
+
+    const String table_metadata_path = getObjectMetadataPath(table_name);
+    const UUID uuid_table = GetTableUUIDFromDetachedMetadata(local_context, table_metadata_path);
+
+    waitDetachedTableNotInUse(uuid_table);
+
+    const StorageID storage_id{getDatabaseName(), table_name, uuid_table};
+    const String table_metadata_path_drop = DatabaseCatalog::instance().getPathForDroppedMetadata(storage_id);
+
+    {
+        std::lock_guard lock(mutex);
+
+        auto txn = local_context->getZooKeeperMetadataTransaction();
+        if (txn && !local_context->isInternalSubquery())
+            txn->commit();
+
+        fs::rename(table_metadata_path, table_metadata_path_drop);
+        LOG_TRACE(log, "Rename {} to {} for removing.", table_metadata_path, table_metadata_path_drop);
+
+        DatabaseCatalog::instance().removeUUIDMappingFinally(uuid_table);
+        table_name_to_path.erase(table_name);
+    }
+
+    if (fs::exists(GetPathSymlink(table_name)))
+    {
+        LOG_TRACE(log, "Remove symlink for {}", table_name);
+        tryRemoveSymlink(table_name);
+    }
+
+    removeDetachedPermanentlyFlag(local_context, table_name, table_metadata_path, true);
```

Comment:
Let's check if we cleanup permanently-detached-flags for tables that do not exist on startup (otherwise this flag will remain on disk if server crashes before removing it but after removing the metadata file)

Reply:
Im going to move it in enqueueDroppedTableCleanup (we will remove flag after removing metadata file).